File: enforce.py
#!/usr/bin/env python3
import sys
import time
import requests
import argparse

parser = argparse.ArgumentParser(description='Infinity Next Enforce Utility')
parser.add_argument(
    '--region', help='The Infinity Next region to enforce policy on: eu or us (default: eu)', default="eu")
parser.add_argument(
    '--client-id', help='The Infinity Next client ID: required', required=True)
parser.add_argument(
    '--access-key', help='The Infinity Next access key: required', required=True)
args = parser.parse_args()

if args.region == "eu":
    inext_url = "https://cloudinfra-gw.portal.checkpoint.com"
elif args.region == "us":
    inext_url = "https://cloudinfra-gw-us.portal.checkpoint.com"
else:
    raise argparse.ArgumentError(argument='region', message='got unknown region ' + args.region + ', expected eu or us')

print('authenticating to ' + inext_url)

auth_reponse = requests.post(inext_url+'/auth/external', headers={'Content-Type': 'application/x-www-form-urlencoded'}, data={
                             'clientId': args.client_id, 'accessKey': args.access_key})
if auth_reponse.status_code != 200:
    raise SystemExit('failed authenticating, got response ' + str(auth_reponse.content) + ' exiting...')
auth = auth_reponse.json()

if 'data' not in auth.keys() or 'token' not in auth['data'].keys():
    raise SystemExit('got malformed authentication response ' + str(auth) + ' exiting...')
token = auth['data']['token']

enforce_reponse = requests.post(inext_url + '/app/i2/graphql/V1', headers={'Authorization': 'Bearer ' + token, 'Content-Type': 'application/json'}, json={
                                "query": "mutation {enforcePolicy {id}}", "variables": {}})
if enforce_reponse.status_code != 200:
    raise SystemExit('failed enforcing policy, got response ' + str(enforce_reponse.content) + ' exiting...')
enforce = enforce_reponse.json()
if 'data' not in enforce.keys() or 'enforcePolicy' not in enforce['data'].keys() or 'id' not in enforce['data']['enforcePolicy'].keys():
    raise SystemExit('got malformed enforce policy response ' + str(enforce) + ' exiting...')
task_id = enforce['data']['enforcePolicy']['id']

task_status = 'InProgress'
print('Waiting for task ' + task_id + ' to complete...')
while task_status == 'InProgress':
    task_response = requests.post(inext_url + '/app/i2/graphql/V1', headers={'Authorization': 'Bearer ' + token, 'Content-Type': 'application/json'}, json={
        "query": "query {getTask(id: \""+task_id+"\") {id\r\nstatus}}", "variables": {}})
    if task_response.status_code != 200:
        print('failed polling task ' + task_id + ' status with response: ' + task_response.content)

    task = task_response.json()
    if 'data' not in task.keys() or 'getTask' not in task['data'].keys() or 'status' not in task['data']['getTask'].keys():
        raise SystemExit('got malformed task response ' + str(task) + ' exiting...')
    task_status = task['data']['getTask']['status']
    time.sleep(0.5)

print('Enforce policy task ' + task_id + ' done with status "' + task_status+'"')
sys.exit(0)

File: cmd/enforce.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/CheckPointSW/infinity-next-cli/utils"
	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
)

type enforcePolicyResponseEnforcePolicy struct {
	ID string `json:"id"`
}

type enforcePolicyResponseData struct {
	EnforcePolicy enforcePolicyResponseEnforcePolicy `json:"enforcePolicy"`
}

// enforceCmd represents the enforce command
var enforceCmd = &cobra.Command{
	Use:   "enforce",
	Short: "Enforce a policy",
	Long:  `Enforce a policy`,

	PreRunE: func(cmd *cobra.Command, args []string) error {
		cmd.Flags().StringVarP(&clientID, "client-id", "c", "", "Client ID of the API key")
		cmd.Flags().StringVarP(&accessKey, "access-key", "k", "", "Access key of the API key")
		cmd.Flags().StringVarP(&region, "region", "r", "eu", "Region of Infinity Next API")
		cmd.Flags().StringVarP(&token, "token", "t", "", "Authorization token of the API key")

		if err := viper.BindPFlags(cmd.Flags()); err != nil {
			return err
		}

		cmd.Flags().VisitAll(func(f *pflag.Flag) {
			if viper.IsSet(f.Name) && viper.GetString(f.Name) != "" {
				if err := cmd.Flags().Set(f.Name, viper.GetString(f.Name)); err != nil {
					fmt.Println(err)
				}
			}
		})

		if err := cmd.MarkFlagRequired("client-id"); err != nil {
			return err
		}

		if err := cmd.MarkFlagRequired("access-key"); err != nil {
			return err
		}

		return nil
	},
	RunE: func(cmd *cobra.Command, args []string) error {
		var URL string
		switch region {
		case "eu":
			URL = EUCIURL
		case "us":
			URL = USCIURL
		default:
			fmt.Printf("Invalid region %s, expected eu or us\n", region)
			os.Exit(1)
		}

		authForm := url.Values{}
		authForm.Add("clientId", clientID)
		authForm.Add("accessKey", accessKey)
		authReq, err := http.NewRequest(http.MethodPost, URL+CIAuthPath, strings.NewReader(authForm.Encode()))
		if err != nil {
			return err
		}

		authReq.Header.Add("Content-Type", "application/x-www-form-urlencoded")
		client := http.Client{}
		authResp, err := client.Do(authReq)
		if err != nil {
			return err
		}

		if authResp.StatusCode != http.StatusOK {
			return fmt.Errorf("failed authenticating to Infinity Next with status %s\n", authResp.Status)
		}

		bResp, err := io.ReadAll(authResp.Body)
		if err != nil {
			return err
		}

		var auth externalAuthResponse
		if err := json.Unmarshal(bResp, &auth); err != nil {
			return err
		}

		var graphqlReq graphqlRequest
		graphqlReq.Query = `mutation {enforcePolicy {id}}`
		graphqlReq.Variables = map[string]interface{}{}

		bReq, err := json.Marshal(graphqlReq)
		if err != nil {
			return err
		}

		enforceReq, err := http.NewRequest(http.MethodPost, URL+CIAPIV1, bytes.NewBuffer(bReq))
		if err != nil {
			return err
		}

		enforceReq.Header.Set("Authorization", "Bearer "+auth.Data.Token)
		enforceReq.Header.Set("Content-Type", "application/json")

		var enforcePolicy graphqlResponse[enforcePolicyResponseData]
		enforceResp, err := utils.HTTPRequestUnmarshal(&client, enforceReq, &enforcePolicy)
		if err != nil {
			return err
		}

		if enforceResp.StatusCode != http.StatusOK {
			return fmt.Errorf("failed enforcing policy with status %s\n", enforceResp.Status)
		}

		taskStatus := "InProgress"
		graphqlReq.Query = "query {getTask(id: \"" + enforcePolicy.Data.EnforcePolicy.ID + "\") {id\r\nstatus}}"
		bReq, err = json.Marshal(graphqlReq)
		if err != nil {
			return err
		}

		errch := make(chan error, 1)
		go func() {
			// Poll for the enforce policy task's status until it's done
			for taskStatus == "InProgress" {
				taskReq, err := http.NewRequest(http.MethodPost, URL+CIAPIV1, bytes.NewBuffer(bReq))
				if err != nil {
					errch <- err
				}

				taskReq.Header.Set("Authorization", "Bearer "+auth.Data.Token)
				taskReq.Header.Set("Content-Type", "application/json")

				var getTask getTaskResponse
				taskResp, err := utils.HTTPRequestUnmarshal(&client, taskReq, &getTask)
				if err != nil {
					errch <- err
				}

				if taskResp.StatusCode != http.StatusOK {
					errch <- fmt.Errorf("failed to get task %s with status %s\n", enforcePolicy.Data.EnforcePolicy.ID, enforceResp.Status)
				}

				taskStatus = getTask.Data.GetTask.Status
				time.Sleep(200 * time.Millisecond)
			}

			errch <- nil
		}()

		// Timeout of 10 seconds for the polling routine to finish
		select {
		case err := <-errch:
			if err != nil {
				return err
			}

			switch taskStatus {
			case "Succeeded":
				fmt.Printf("Enforce policy task %s succeeded\n", enforcePolicy.Data.EnforcePolicy.ID)
			case "Failed":
				return fmt.Errorf("enforce policy task %s failed", enforcePolicy.Data.EnforcePolicy.ID)
			default:
				return fmt.Errorf("enforce policy task %s done with unknown status %s", enforcePolicy.Data.EnforcePolicy.ID, taskStatus)
			}
		case <-time.After(10 * time.Second):
			return fmt.Errorf("enforce policy task did not finish after 10 seconds, quiting")
		}

		return nil
	},
}

func init() {
	rootCmd.AddCommand(enforceCmd)
}
